R/pairwise-comparisons.R

```diff
@@ -0,0 +1,244 @@
+
+    # ratio should be equivalent to theta_ij in Johannes document
+    ratio <- sum(values_x) / sum(values_y)
+
+    if (permutation_test) {
+      # taken from the surveillance package
+      permutation_test <- function(score1, score2, nPermutation = 999, ratio = FALSE) {
+        nTime = length(score1)
+        meanScore1 <- mean(score1)
+        meanScore2 <- mean(score2)
+
+        if (ratio) {
+          testStat_observed <- meanScore1 / meanScore2
+        } else {
+          testStat_observed <- meanScore1 - meanScore2
+        }
+
+        testStat_permuted <- replicate(nPermutation, {
+          sel <- rbinom(nTime, size = 1, prob = 0.5)
+          g1 <- (sum(score1[sel == 0]) + sum(score2[sel == 1]))/nTime
+          g2 <- (sum(score1[sel == 1]) + sum(score2[sel == 0]))/nTime
+
+          if (ratio) {
+            g1 / g2
+          } else {
+            g1 - g2
+          }
+        })
+
+        pVal <- (1 + sum(abs(testStat_permuted) >= abs(testStat_observed)))/(nPermutation + 1)
+        # plus ones to make sure p-val is never 0?
+        return(pVal)
+      }
+
+      pval <- permutation_test(values_x, values_y)
+    } else {
+      pval <- NA
+    }
+    return(list(ratio = ratio, pval = pval))
+  }
+
+  # we only need to do the calculation once, because for the ratio that
+  # should just be the inverse and for the permuation the result should
+  # be the same
+
+  combinations <-   data.table::as.data.table(t(combn(identifiers, m = 2)))
+  colnames(combinations) <- c("model", "compare_against")
+
+  combinations[, c("ratio", "pval") := single_comparison(model, compare_against,
+                                                         permutation_test),
+               by = seq_len(NROW(combinations))]
+  combinations[sort(ratio)]
+
+  # mirror computations
+  combinations2 <- data.table::copy(combinations)
+  data.table::setnames(combinations2, old = c("model", "compare_against"),
+                       new = c("compare_against", "model"))
+  combinations2[, ratio := 1/ratio]
+
+  # add a one for those that are the same
+  combinations3 <- data.table::data.table(model = identifiers,
+                                          compare_against = identifiers,
+                                          ratio = 1,
+                                          pval = NA)
+
+  result <- data.table::rbindlist(list(combinations,
+                                       combinations2,
+                                       combinations3),
+                                  use.names = TRUE)
+
+  # calculate relative wis as geometric mean
+  # need a different name for the variable! mean_ratio?
+  result[, theta_i := exp(mean(log(ratio))),
+         by = "model"]
+
+  if(!is.null(baseline)) {
+    baseline_theta <- unique(result[model == baseline, ]$theta_i)
+    result[, rel_to_baseline := theta_i / baseline_theta]
+  }
+
+  return(result)
+}
+
+
+
+
+
+
+#' @title Plot Heatmap of Pairwise Comparisons
```

Comment:
is this only designed to take the output of the function above and nothing else? If yes could use an S3 class. Again been slightly changing my view on these as I like that they reduce writing and memory overhead. Downside is that they are then keyed to the class of the input which isn't ideal.

Reply:
hm. never been using any S3 classes....I think it could in principle also take other things as input - I just wrote it that way to clarify what the expected input was

Reply:
They are pretty easy and might be nice to suppor t alongside your named functions. See the estimate_secondary.R script in EpiNow2 for a use case.

Reply:
I'll have a look but will probably not implement it right now

Reply:
👍